### PR TITLE
out_cloudwatch_logs: fix memory leak when using cpu or mem input

### LINKED ISSUE #3145: Memory leak in cloudwatch_logs output when mem or cpu inputs are enabled
## Bug Report

**Describe the bug**
A memory leak occurs when the `cloudwatch_logs` output plugin processes input from the `cpu` or `mem` input plugins. `td-agent-bit` memory usage increases over time until the process is killed or restarted.

**To Reproduce**
- Example configuration (obviously change `region`, `log_group_name` & `log_stream_prefix` as required):
```
[SERVICE]
    flush        1
    daemon       Off
    log_level    info
    parsers_file parsers.conf
    plugins_file plugins.conf
    http_server  Off

[INPUT]
    name cpu
    Tag  cpu

[INPUT]
    Name   mem
    Tag    memory

[OUTPUT]
    Name              cloudwatch_logs
    Match             *
    region            eu-west-2
    log_group_name    fluent-bit-example
    log_stream_prefix memory-leak-example

```
- Ensure this is running somewhere with the necessary IAM permissions to write to CloudWatch

**Expected behavior**
`td-agent-bit` memory usage should be fairly stable over time.
Instead, memory usage increases without limit. In our case, memory usage reached 8GB in less than 1 week

Note that with the cpu/memory inputs enabled, but with a _different output plugin_, the issue does not occur, also with the cloudwatch output plugin enabled, but with _different input plugins_, the issue does not occur.

**Your Environment**

* Version used:
```
$ /opt/td-agent-bit/bin/td-agent-bit --version
Fluent Bit v1.7.0
```
* Configuration:
    * See example config file above
* Server type and version:
* Operating System and version:
    * Various EC2 instance types running a mixture of Debian & Red Hat
* Filters and plugins:
    * See example config file above

**Additional context**
Memory leaks were possibly identified using valgrind `valgrind --leak-check=full --show-leak-kinds=all --track-origins=yes /opt/td-agent-bit/bin/td-agent-bit -c /etc/td-agent-bit/td-agent-bit.conf`
The relevant blocks i can find are
```
==241309== 960 (192 direct, 768 indirect) bytes in 2 blocks are definitely lost in loss record 646 of 677
==241309==    at 0x4C30F0B: malloc (vg_replace_malloc.c:307)
==241309==    by 0x569A3C: flb_malloc (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x56C3CA: process_and_send (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x569547: cb_cloudwatch_flush (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x4BF60A: output_pre_cb_flush (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x902A85: co_init (in /opt/td-agent-bit/bin/td-agent-bit)
```
Which I believe is https://github.com/fluent/fluent-bit/blob/master/plugins/out_cloudwatch_logs/cloudwatch_api.c#L891 - where metric is malloc’d inside a for loop without being free’d, though there is a free outside the for loop at line 909.

```
==241309== 20,496 (4,112 direct, 16,384 indirect) bytes in 2 blocks are definitely lost in loss record 675 of 677
==241309==    at 0x4C30F0B: malloc (vg_replace_malloc.c:307)
==241309==    by 0x8E4FE8: init_chunk_list (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x8E5481: msgpack_zone_init (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x56BF61: pack_emf_payload (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x56C4DE: process_and_send (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x569547: cb_cloudwatch_flush (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x4BF60A: output_pre_cb_flush (in /opt/td-agent-bit/bin/td-agent-bit)
==241309==    by 0x902A85: co_init (in /opt/td-agent-bit/bin/td-agent-bit)
```
Which points to https://github.com/fluent/fluent-bit/blob/master/plugins/out_cloudwatch_logs/cloudwatch_api.c#L767 - where there is a msgpack_zone_init without a corresponding msgpack_zone_destroy. And at https://github.com/fluent/fluent-bit/blob/master/plugins/out_cloudwatch_logs/cloudwatch_api.c#L665 there is a msgpack_bufer_init without a msgpack_sbuffer_destroy

### PR DESCRIPTION
**Pull request by adefisher**
Signed-off-by: Ade Fisher adrian.fisher@montvieux.com

Fixes memory leak in out_cloudwatch_logs when using cpu or mem input

Fixes #3145 

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

----
Valgrind output after change

```
[2021/03/31 08:52:04] [ info] [engine] service stopped
==85211== 
==85211== HEAP SUMMARY:
==85211==     in use at exit: 0 bytes in 0 blocks
==85211==   total heap usage: 193,370 allocs, 193,370 frees, 33,548,850 bytes allocated
==85211== 
==85211== All heap blocks were freed -- no leaks are possible
==85211== 
==85211== For lists of detected and suppressed errors, rerun with: -s
==85211== ERROR SUMMARY: 23679 errors from 132 contexts (suppressed: 0 from 0)
```

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by edsiper** [comment]
assigned to @PettitWesley

**Comment by PettitWesley** [comment]
@hossain-rayhan This is your code- can you test it and review?

File: plugins/out_cloudwatch_logs/cloudwatch_api.c [hunk 1]
```diff
@@ -798,6 +802,9 @@ int process_and_send(struct flb_cloudwatch *ctx, const char *input_plugin,
     int check = FLB_FALSE;
     int found = FLB_FALSE;
     struct flb_time tms;
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct flb_intermediate_metric *an_item;
```
Comment (hossain-rayhan) [inline plugins/out_cloudwatch_logs/cloudwatch_api.c:807]: nit: can we put these after line 810? The comment at line 809, will make more sense.

**Comment by hossain-rayhan** [comment]
I tested the plugin and everything is working fine. However, the result from valgrind is little bit different for me. I think we have some memory issues but these are not from current changes. I ran the follwoing command.

**Command**
`valgrind --leak-check=yes --track-origins=yes bin/fluent-bit -c fluent-bit.conf`. 

**Output**

Screen Shot 2021-04-09 at 10 22 55 AM

**Review by hossain-rayhan — DISMISSED** [review]

**Comment by PettitWesley** [comment]
@adefisher Can you please squash all of your commits into one commit? Then I can approve and merge this.

**Comment by adefisher** [comment]
@PettitWesley commits have now been squashed

**Review by PettitWesley — APPROVED** [review]

**Comment by PettitWesley** [comment]
@adefisher Thanks! Will you please also cherry pick this change onto the 1.7 branch and submit a PR against it? That way we can get this fix out in the 1.7.x series (master is only used for new versions- so the PR in this commit won't go out till 1.8.0).

**Pull request merged**